zksync-era integration docs (#211)
lambdaclass/era_vm: docs/zksync-era-integration.md

## Zksync-era integration

> The `era_vm` doesn't do much by itself, it needs someone to tell it what to do and orquestrate it. On ethereum, for example, the execution node would take a block and process all of the transactions within it one by one using the `evm`. In zksync, things are a bit different, here transactions aren't processed one by one but in batch, this is to publish all of the transactions as just a single one, making the l1 post cheaper and distribitued among all the trnansactions withing the block.

Reviewer: ```suggestion
The `era_vm` doesn't do much by itself, it needs someone to tell it what to do and orchestrate it. On Ethereum, for example, the execution node would take a block and process all of the transactions within it one by one using the `evm`. In zksync, things are a bit different, here transactions aren't processed one by one but in batch, this is to publish all of the transactions as just a single one, making the l1 post cheaper and distributed among all the transactions within the block.
```

This is what the bootloader do. The bootloader is a `.yul` contract that lives on l1 and takes an array of transactions and executes them in the `era_vm`.

> Part of integrating our vm with the zk stack comes down to integrate the bootloader with our vm. That is, we need to make our vm "talk" with the bootloader. Before diving into the integration, we'll first explore what the bootloader does more in detail, which will explain what parts we need to keep track on our vm.

Reviewer: > Part of integrating our vm with the zk stack comes down to integrate the bootloader with our vm. That is, we need to make our vm "talk" with the bootloader.

Couldn't the bootloader just run on top of our VM? I believe the work on the bootloader is more about giving us some flexibility and removing overhead for a "fast run" VM for which a proof is not expected.

Author: Btw, there a lot of typos there I am aware I need to fix.

# ZK stack

## Introduction

> So far we have been talking only about the `era_vm`. But you should know that the vm is only a small part of the zk stack. The zk stack is composed of many critical components. In this section, we are only going to be interested in one particular component: the **nodes**, which can further be decomposed into the following units:

Reviewer 2: > we are only going to be interested in one particular component: the **nodes**, which can further be decomposed into the following units:

The word *node* is a bit overloaded, I would rephrase this as saying the component we are interested in is the **Operator/Sequencer**, explain what it is and talk about both the bootloader and the VM that it uses

> -   **Operator/Sequencer**: the server that initializes the vm, injects the Bootloader bytecode, receives transactions and pushes them into the vm(bootloader memory to be more specific) and start batches and seals them.

Reviewer 3: ```suggestion
-   **Operator/Sequencer**: the server that initializes the vm, injects the Bootloader bytecode, receives transactions and pushes them into the vm (bootloader memory to be more specific) and start batches and seals them.
```

-   **Bootloader**: a system contract that receives an array of transactions which are processed, validated, executed, and then, the final state is published in the l1.
-   **era_vm**: the virtual machine where the Bootloader(and so all the transactions bytecode) gets executed.

These components interact continuously to process transactions. This document will provide an overview of the Bootloader, then explore the operator, its management of the bootloader, and finally, the data publishing process to L1. All of this while primarily focusing on how these interactions impact the design of the `era_vm`.

## Bootloader

The Bootloader is a special system contract whose hash resides on L1, but its code isn't stored on either L1 or L2. Instead, it’s compiled from `.yul` to `era_vm` assembly using `zksolc` when the operator first initializes the VM (more on that below).

> The Bootloader, unlike Ethereum, takes an array of transactions(a batch) and executes all of them in one run (unless specified not to, that is, if the execution mode of the vm is set to OneTx). This approach allows the transaction batch to be posted on the l1 as just a single one, making the processing on Ethereum cheaper, since taxes and gas can be distributed among all the transactions within the posted batch.

Reviewer 3: ```suggestion
The Bootloader, unlike Ethereum, takes an array of transactions (a batch) and executes all of them in one run (unless specified not to, that is, if the execution mode of the vm is set to OneTx). This approach allows the transaction batch to be posted on the l1 as just a single one, making the processing on Ethereum cheaper, since taxes and gas can be distributed among all the transactions within the posted batch.
```

At the most basic level, the Bootloader performs the following steps:

1. Reads the initial batch information and makes a call to the SystemContext contract to validate the batch.

> 2. Loops through all transactions and executes them until the `execute` flag is set to $0$, at that point, it jumps to step 3. <!-- Here we could also add that the transaction gets processed based on where it came from (l1 or l2) -->

Reviewer 3: ```suggestion
2. Loops through all transactions and executes them until the `execute` flag is set to `0`, at that point, it jumps to step `3`. <!-- Here we could also add that the transaction gets processed based on where it came from (l1 or l2) -->
```

> 2. Loops through all transactions and executes them until the `execute` flag is set to $0$, at that point, it jumps to step 3. <!-- Here we could also add that the transaction gets processed based on where it came from (l1 or l2) -->

Reviewer 2: <!-- Here we could also add that the transaction gets processed based on where it came from (l1 or l2) -->

Let's add a small comment about this, doesn't need to explain how l1/l2 transactions differ from each other but at least mention them

3. Seals l2 block and publish final data to the l1.

The initial validation of the batch is necessary, since, as we'll see below, the Bootloader starts with its memory pre-filled with any data the operator wants. That is why it needs to validate its correctness.

For more details, you can see the [main loop](https://github.com/matter-labs/era-contracts/blob/main/system-contracts/Bootloader/bootloader.yul#L3962-L3965) or the [full contract code](https://github.com/matter-labs/era-contracts/blob/main/system-contracts/bootloader/bootloader.yu).

## Operator/sequencer

> Currently, the operator is a centralized server (there are plans to make a decentralized consensus for operators) which can be thought of as the node entry point, its responsibilities are:

Reviewer 2: > which can be thought of as the node entry point

Again, I would remove this since the word `node` is too overloaded (right now the word `node` and operator` are usually interchangeable actually)

-   Keeping storage database and committing changes.

Here is a simplified version of what the vm on the operator end looks like:

```rust
struct OperatorVm {
    pub(crate) inner: EraVM, // this would be the actual `era_vm`
    pub suspended_at: u16, // last pc when execution stopped because of a hook
    pub Bootloader_state: BootloaderState,
    pub(crate) storage: StorageDb,
    pub snapshot: Option<VmSnapshot>,
}
```

### Initializing the vm

VM initialization involves:

-   Loading the Bootloader bytecode and initializing its state.
-   Setting up the `era_vm` by injecting the Bootloader code, loading default contracts, and configuring other settings.

When setting up `era_vm`, the operator provides access to the chain storage database with the following API:

```rust
trait Storage {
    fn decommit(&mut self, hash: U256) -> Option<Vec<U256>>;

    fn storage_read(&mut self, key: &StorageKey) -> Option<U256>;

    fn cost_of_writing_storage(&mut self, key: &StorageKey, value: U256) -> u32;

    fn is_free_storage_slot(&self, key: &StorageKey) -> bool;
}
```

This storage is saved in the VM state as a pointer. Here’s a brief explanation of each function:

> -   **decommit**: given a hash it returns a contract bytecode from the database.

Reviewer 2: > given a hash it returns a contract bytecode from the database.

given a hash it returns *its corresponding bytecode* (if it exists) from the database

-   **storage_read**: given a key, it returns the potential value from the database.
-   **cost_of_writing_storage**: when writing to the contract storage, gas is consumed, but the cost of writing depends on whether the write is initial or not. More on that [here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/docs/specs/zk_evm/fee_model.md).
-   **is_free_storage_slot**: if the address to write belongs to the system context and the key belongs to the base L2 token address, then the storage_slot is free(doesn't incur gas charges).

A few notes about this storage:

1. The key has the following structure:

```rust
struct StorageKey {
    pub address: H160,
    pub key: U256,
}
```

> This allows us to query a key that belongs to the current executing address.

Reviewer 2: > This allows us to query a key that belongs to the current executing address.

This allows us to query a storage key belonging to any desired contract through its address

> 2. There isn't any consensus or spec about how storage should be implemented. We came up with this API because it is what we thought was more convenient for the requirement. But, for example, the vm1 implements a query logic, where the operator will react based on the provided params.

Reviewer 3: ```suggestion
2. There isn't any consensus or spec about how storage should be implemented. We came up with this API because it is what we thought was more convenient for the requirement. But, for example, the vm1 implements a query logic, where the operator will react based on the [provided params](https://github.com/matter-labs/zksync-era/blob/87768755e8653e4be5f29945b56fd05a5246d5a8/core/lib/types/src/zk_evm_types.rs#L17-L30).
```

> 2. There isn't any consensus or spec about how storage should be implemented. We came up with this API because it is what we thought was more convenient for the requirement. But, for example, the vm1 implements a query logic, where the operator will react based on the provided params.

Reviewer 2: > But, for example, the vm1 implements a query logic, where the operator will react based on the provided params.

Can we give a concrete example of what this looks like?

> [Here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/core/lib/multivm/src/versions/era_vm/vm.rs#L726) you can take a look a the implementation of this trait in detail.

Reviewer 3: ```suggestion
[Here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/core/lib/multivm/src/versions/era_vm/vm.rs#L726) you can take a look at the implementation of this trait in detail.
```

> [Here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/core/lib/multivm/src/versions/era_vm/vm.rs#L726) you can take a look a the implementation of this trait in detail.

Reviewer 3: We should probably give a brief explanation about our `World<S>` before showing something it implements.

-   **PostResult**: Sets the last transaction result
-   **TxHasEnded**: if the mode of execution is set to **OneTx**, then the execution is stopped and it returns the result collected in the _PostResult_ hook.
-   **NotifyAboutRefunds**: Inform the operator about the amount of gas refunded after a transaction.
-   **AskOperatorForRefund**: here the Bootloader asks the operator to suggest a refund amount for the transaction.
-   **PubdataRequested**: At the end of the batch, the Bootloader asks for the data to publish on the l1 (more on this later).

Now, where does the operator know where to write to? Again, within the `era_vm`, there exists a special heap reserved exclusively for the Bootloader. The Operator writes all the data in that heap which has designated slots based on the type of data to write (see more [here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/docs/specs/zk_evm/bootloader.md#structure-of-the-bootloaders-memory)). Transactions, for example, are pushed into the `[252189..523261]` slots.

### Rollbacks and snapshots

> In the `era_vm`, when a transaction encounters a panic or reverts, the vm needs to roll back the changes, restoring only a part of the [state](https://github.com/lambdaclass/era_vm/blob/main/src/state.rs#L43-L60) to its previous frame. Remember that frames are created under `near_call` and `far_call` opcodes. Currently, rollbacks are perform using snapshots which are just copies of the current state. If a rollback is necessary, the state is restored from these snapshots.

Reviewer 3: This section may be better suited [here](https://github.com/lambdaclass/era_vm/blob/main/docs/general.md#difference-between-a-revert-and-a-panic-exception-handlers).

Author: The idea was to recall the idea of rollbacks in `era_vm` to emphasize on the difference  with bootloader rollbacks.

> The Bootloader, can fail sometimes, and it is the job of the Operator to trigger rollbacks. However, this type of rollback differs from the ones just mentioned. Bootloader rollbacks involve restoring not only the [full vm state](https://github.com/lambdaclass/era_vm/blob/zksync-era-integration-tests/src/vm.rs#L66-L69) but also the bootloader state. These snapshots are called external snapshots and can only be triggered by the Bootloader. [Here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/core/lib/multivm/src/versions/era_vm/snapshot.rs) you can see what a full snapshot looks like. Before starting a new batch execution, the operator creates a snapshot, which is also used at the end of execution to collect logs (see [here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/core/lib/multivm/src/versions/era_vm/vm.rs#L508-L595) for more details).

Reviewer 3: ```suggestion
The Bootloader may fail sometimes, it is the job of the Operator to trigger the necessary rollbacks. However, this type of rollback differs from the ones just mentioned. Bootloader rollbacks involve restoring not only the [full vm state](https://github.com/lambdaclass/era_vm/blob/zksync-era-integration-tests/src/vm.rs#L66-L69) but also the bootloader state. These snapshots are called external snapshots and can only be triggered by the Bootloader. [Here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/core/lib/multivm/src/versions/era_vm/snapshot.rs) you can see what a full snapshot looks like. Before starting a new batch execution, the operator creates a snapshot, which is also used at the end of execution to collect logs (see [here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/core/lib/multivm/src/versions/era_vm/vm.rs#L508-L595) for more details).
```

## Publishing data

As said above, once the batch of transactions has all been executed, the final step in the Bootloader is to publish the final data. The data to be published is composed of:

-   **L2 to L1 Logs**: Logs generated during L2 transactions that need to be recorded on L1. This can be transactions on L1 that have been forwarded to the L2 to lower costs.
-   **L2 to L1 Messages**: used to transmit instructions or data from smart contracts on L2 to contracts or systems on L1.
-   **Smart Contract Bytecodes**: This involves publishing the bytecodes of smart contracts deployed on L2. Before being sent to L1, these bytecodes are often compressed to save space and reduce costs.
-   **Storage writes**: These are records of changes to the storage on L2. Only the final diff from the previous state is included.

In theory, with this data one should be able to reconstruct the whole state of the l2.

> At the end of the batch, the Bootloader calls the `PubdataRequested` hook to ask the operator for the final batch state. The operator writes into the bootloader memory(slots [40053..248052]) the collected data from the`era_vm`. [Here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/core/lib/multivm/src/versions/era_vm/vm.rs#L312-L353) you can see the hook implementation in detail.

Reviewer 3: ```suggestion
At the end of the batch, the Bootloader calls the `PubdataRequested` hook to ask the operator for the final batch state. The operator writes into the bootloader memory(slots [40053..248052]) the collected data from the`era_vm`. [Here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/core/lib/multivm/src/versions/era_vm/vm.rs#L309-L350) you can see the hook implementation in detail.
```

<a id="era_vm-key-structures"></a>
And so we end up with two key structures on the `era_vm`:

-   [The execution state](https://github.com/lambdaclass/era_vm/blob/main/src/execution.rs#L32-L51): the state of registers, heaps, frames, etc.
-   The L2 state changes: the changes on the chain that will get published on L1 and committed to the l2 database.

Finally, everything finishes with the operator committing the changes to its database.

### Refunds, Storage write/read and Pubdata Costs associated

We keep several fields in the `VMState` to track refunds and pubdata associated costs. Refunds are a return of ergs spent, since gas is always paid upfront. They can occur during storage operations, specifically when:

-   Reading from storage
-   Writing to storage

The relevant opcodes for these operations are:

-   `far_call`
-   `decommit`
-   `sstore`
-   `ssload`

#### How Refunds Are Calculated

Refunds depend on whether a storage key has been accessed before. To manage this, we use the following keys from the `VMState`:

-   `paid_changes`
-   `read_storage_slots`
-   `written_storage_slots`
-   `decommited_hashes`

#### Decommit behaviour

> Decommits might occur during `far_call` or `decommit`. Whenever we decommit a `hash`, we check if that hash has already been decommited, if it is then we return the gas spent for deommit since decommits are paid upfront. Otherwise, we store the hash a already decommited so subsequent decommits to that hash will become free of charge.

Reviewer 4: There are typos here: `deommit` and `the hash a already decommited`

#### Storage Read Behavior

When performing a storage_read, we check if the slot is free or if the key has already been read. If the key has been read before, a "warm" refund is given. If not, no refund is provided, but the key is marked as read for future refunds.

#### Storage Write Behavior

> During a storage_write, we first check if the slot is free. If it is, a "warm" refund is given. Otherwise, we calculate the pubdata cost—the current price for writing to storage. If the key has been written to before, we only pay the difference between the new price and the previously paid amount (this difference is what we track as `pubdata_costs`). Notice that of this difference is negative (i.e the new price is lower that what has already been paid), then they cost becomes free of charge, but it doesn't actually result in a refund. Additionally, if the key has been written to before, a "warm" refund is provided. If the key has only been read before and is now being written to, a "cold" write refund is given.

Reviewer 4: Typo: `Notice that of this difference`
